### PR TITLE
DOC: Initialize numpy compatibility note

### LINKED ISSUE #48628: Add docs on PyTorch - NumPy interaction
Mixing PyTorch and NumPy is quite common and there are a lot of subtleties involved (see for example gh-7610), but there is no documentation on this at all right now. I think this belongs as a separate section in the API docs.

_Outline:_

Converting tensors into ndarrays, and vice versa. Should cover:
- Constructor functions (`tensor`, `as_tensor`, `asarray`)
- `Tensor.numpy()`
- The purpose of `__array__`, when it works, how to deal with it when it doesn't (see gh-36560)
- When conversion is or isn't zero-copy, and that if it is memory is shared and mutations affect both tensor and array

Using NumPy functions directly on PyTorch tensors:
- When that works
- When a tensor is returned (this is designed behaviour, see `Tensor.__array_wrap__`) and when an `ndarray` or scalar is produced
- Mention that the reverse, applying PyTorch functions on NumPy arrays, never works (and the rationale for that)
- Reductions have special-case behavior in NumPy (e.g. `np.max[tensor]`) will call `tensor.max()`).

Python operators applied to a tensor and an array:
- `tensor + array` et al., precedence rules and what type gets returned (see gh-46829)

Other places where tensors and arrays can be used interchangeably:
- indexing work both ways, `array[tensor]` and `tensor[array]`, for boolean and fancy indexing

Link to what seems to be [the one relevant tutorials](https://pytorch.org/tutorials/advanced/numpy_extensions_tutorial.html), and check that that uses best-practice code.

Also related: `__cuda_array_interface__` is implemented but not documented, mention that it allows zero-copy interop with CuPy, Numba et al.

### PR DESCRIPTION
**Pull request by HaoZeke**
Fixes #48628, #46829. 

- [x] This will need to be revised when #59790 is merged in. 
-  [x] Will also need a minor adjustment (describing the order in which NumPy attempts to convert a foreign object, and the addition of a hyperlink to the NumPy interoperability doc) when https://github.com/numpy/numpy/pull/20185 is merged.

@rgommers

### CODE REVIEW

File: docs/source/notes/numpy_compatibility.rst [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            return torch_tensor.numpy()
+        else:
+            warnings.warn("Calling detach, probably lossy!")
+            return torch_tensor.detach().numpy()
+
+Operations
+------------
+
+All ``torch`` operators will helpfully fail with a ``TypeError`` if called with
+``numpy`` arrays. However, for **numpy methods**, using a ``torch.Tensor`` with
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:70]: methods -> functions?

File: docs/source/notes/numpy_compatibility.rst [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+   import torch
+   import numpy as np
+   a_np = np.ones(3) # dtype=float64
+   b_torch = torch.ones(3) # dtype=float32
+   torch.add(a_np, b_torch) # TypeError
+   b_torch + a_np # TypeError
+   a_np + b_torch # OK but awkward
+   c_tmp = np.add(a_np, b_torch) # OK
+   # c_tmp.dtype is torch.float64
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:91]: To give a concrete example of the doctest format:
```python
>>> c = np.add(a_np, b_torch)
>>> c.dtype
torch.float64
```
Much more readable than these comments.

File: docs/source/notes/numpy_compatibility.rst [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+Since the ``torch.Tensor`` methods will not work with ``np.ndarray`` objects, to
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:93]: methods -> functions

File: docs/source/notes/numpy_compatibility.rst (hunk 2 above)
Comment (rgommers) [reply docs/source/notes/numpy_compatibility.rst:91]: _You want to show the example, and then explain why that behaves as it does I think. Here is what I would write:_

```python
>>> x + t
...
TypeError: Concatenation operation is not implemented for NumPy arrays, use np.concatenate() instead. Please do not rely on this error; it may not be given on all Python implementations.

>>> t + x
tensor([4., 4., 4.], dtype=torch.float64)
```
The reason that `x + t` errors here is that, due to the rules of the Python language for operator support, PyTorch is called here. To be exact, `Tensor.__add__` is called - and that does not accept NumPy arrays.

`t + x` on the other hand calls `ndarray.__add__`. NumPy will then convert the tensor `t` to a `numpy.ndarray` (a zero-copy operation), calculate the result for adding two arrays, and then convert the output back to a `torch.Tensor` (this last step happens because `Tensor` implements an `__array_wrap__` method, instructing NumPy to do that conversion).

File: docs/source/notes/numpy_compatibility.rst [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+prevent the apparent asymmetry of ``a_np * b_torch`` passing while ``b_torch *
+a_np`` fails, the user would **do best** to use NumPy functions instead:
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:95]: better not to recommend this.

File: docs/source/notes/numpy_compatibility.rst [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+    "``bool_``", "``bool``", "N/A"
+    "``uint$``", "``uint$``", ":math:`8`"
+    "``int$``", "``int$``", ":math:`8, 16, 32, 64`"
+    "``float$``", "``float$``", ":math:`16, 32, 64`"
+    "``complex$``", "``complex$``", ":math:`64, 128`"
+
+.. warning::
+
+   The promotion and casting rules `of NumPy`_ differ from those `of PyTorch`_.
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:122]: This warning is not all that useful unless you explain how. What does the reader need to know here?

File: docs/source/notes/numpy_compatibility.rst [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+   # b_torch = torch.as_tensor(a_np) # see note
+   b_torch[2] = 23
+   assert(b_torch.numpy() == a_np) # True
+
+- :meth:`torch.from_numpy()` is guaranteed to share memory with NumPy.
+- :meth:`torch.as_tensor()` will try to stay away from copy operations, it
+  also has the effect of sharing memory. However, ``torch.as_tensor()`` has
+  slightly higher overhead as it checks and accepts other iteratable objects as
+  well, e.g. ``list`` objects.
+- :meth:`torch.from_dlpack()` called with a NumPy array as its argument will also generate a ``torch.Tensor`` view.
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:170]: Please add a note that this requires `numpy >= 1.22.0`

File: docs/source/notes/numpy_compatibility.rst [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+.. note::
+
+   This is due to the fact that ``torch.Tensor.__array_priority__`` is higher
+   than the NumPy default of ``0``.
+
+Indexing
+^^^^^^^^^^^^
+
+For the most part, both simple and "fancy" indexing work as expected. Edge cases
+are enumerated here.
```
Comment (rgommers) [inline docs/source/notes/numpy_compatibility.rst:203]: The main point here was that one can index a tensor with an ndarray, and vice versa. Can you state that, with a few examples? And again a recommendation that explicit is better than implicit - work in one library as a user and do the conversions up front.

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @HaoZeke, this is a good start.

A section _"Using NumPy functions directly on PyTorch tensors"_ still seems to be missing. (concrete example: `np.sin(t)` works. let's explain why). The reductions special-case mentioned in gh-48628 is also missing.

Also note https://github.com/pytorch/pytorch/issues/46829#issuecomment-728166806. Here you use `+` as the operator, but it'd be good to copy that example and mention that numpy may behave unexpectedly for different operators.

File: docs/source/notes/numpy_compatibility.rst [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+NumPy Compatibility
+=====================
+
+.. note::
+
+   **TL;DR:**
+
+   - Be explicit about obtaining and using NumPy arrays, or
+   - Avoid operators (``-``, ``+``, ``*``, ``/``) in favor of methods
+     (``np.subtract``, ``np.add``, ``np.multiply``, ``np.divide``)
```
[the remark this answers is not in the document]
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:12]: The idea was that since `torch` operators don't work with ndarrays, if one would like to not worry about the asymmetry of `a+x` vs `x+a` where `a->np.ndarray` and `x->torch.tensor`, using the methods might be a viable option, but I think it'd be best to drop it.

File: docs/source/notes/numpy_compatibility.rst (hunk 5 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:122]: I think the only point here is that not all the types are supported or have equivalents, but yeah, it is spurious here.

File: docs/source/notes/numpy_compatibility.rst [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+   b_torch.numpy() # Fine, will work
+
+The interoperability guidelines places a strong emphasis on correctness over
+convenience. As the conversion to a NumPy array may potentially be a lossy
+operation (the computational graph), there are no inbuilt operations to do
+something which unconditionally returns a NumPy array.
+
+.. code-block:: python
+
+   # Never do this
```
[the remark this answers is not in the document]
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:55]: Good point, rewording.

File: docs/source/notes/numpy_compatibility.rst (hunk 7 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst]: Yup, gave it a shot.

File: docs/source/notes/numpy_compatibility.rst [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+
+From the end-user perspective, it is most desirable to have zero copy methods,
+however, for exploratory analyses and their visualizations in particular, it is
+often more important to have guaranteed conversions, even if copies are created.
+
+Basic interoperability
+----------------------
+
+NumPy arrays have no conception of the GPU, and so all conversions require a
+``cpu`` tensor. Similarly, the **computational graph** cannot be represented in
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:39]: "Computational graph" is probably too vague and ill-defined a term for most readers

File: docs/source/notes/numpy_compatibility.rst [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+
+.. doctest::
+
+   >>> b_torch = torch.ones(3)
+   >>> assert(b_torch.requires_grad == False)
+   >>> assert(b_torch.is_cuda == False)
+   >>> b_torch.numpy()
+   array([1., 1., 1.], dtype=float32)
+
+PyTorch will raise exceptions instead of allowing operations that may lead to
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:51]: This sentence is very general and I'm not sure it's strictly true?

File: docs/source/notes/numpy_compatibility.rst [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+incorrect results. As the conversion to a NumPy array may potentially be a lossy
+operation (the computational graph), there are no inbuilt operations to do
+something which unconditionally returns a NumPy array. Practically speaking, a
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:54]: Not really sure what this paragraph is trying to say

File: docs/source/notes/numpy_compatibility.rst [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+function can be created locally to return a NumPy array, for say, visualizations.
+
+.. code-block:: python
+
+   # Possibly lossy function for visualizations
+   def uncond_numpy(torch_tensor):
+        if torch_tensor.is_cuda:
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:61]: Tensors can be on other accelerators other than the CPU. A more general check would be if the tensor isn't on the CPU

File: docs/source/notes/numpy_compatibility.rst [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+           torch_tensor = torch_tensor.cpu()
+        if torch_tensor.requires_grad == False:
+            return torch_tensor.numpy()
+        else:
+            return torch_tensor.detach().numpy()
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:66]: Why not unconditionally call `detach()`? 

Don't we sometimes have to call more than `detach()`, like resolving the conj bit?

File: docs/source/notes/numpy_compatibility.rst [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+- **What** is the underlying data?
+    * In particular, several functions will either return a copy or a view
+- **How** are modifications to the data handled?
+    * The data returned may be read-only, or may be writable
+
+From the end-user perspective, it is most desirable to have zero copy methods,
+however, for exploratory analyses and their visualizations in particular, it is
+often more important to have guaranteed conversions, even if copies are created.
+
+Basic interoperability
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:35]: How and when tensors share memory is interesting, too. Should this cover converting NumPy arrays to torch tensors and converting torch tensors to NumPy arrays?

File: docs/source/notes/numpy_compatibility.rst [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+           torch_tensor = torch_tensor.cpu()
+        if torch_tensor.requires_grad == False:
+            return torch_tensor.numpy()
+        else:
+            return torch_tensor.detach().numpy()
+
+Operations
+----------
+
+All ``torch`` operators will helpfully fail with a ``TypeError`` if called with
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:71]: I'm not sure this is true

File: docs/source/notes/numpy_compatibility.rst [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+``numpy`` arrays. However, for **numpy functions**, using a ``torch.Tensor`` with
+an ``np.ndarray`` will return a ``torch.Tensor`` without creating any additional
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:73]: This is unfortunately not true

File: docs/source/notes/numpy_compatibility.rst [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+   >>> a_np + b_torch
+   Traceback (most recent call last):
+   ...
+   TypeError: Concatenation operation is not implemented for NumPy arrays, use np.concatenate() instead. Please do not rely on this error; it may not be given on all Python implementations.
+   >>> np.add(a_np, b_torch)
+   tensor([2., 2., 2.], dtype=torch.float64)
+
+To recognize the root cause of the errors and the legitimacy of the results above, we will recall that:
+
+- The `Python data model`_ specifies that the ``__radd__`` function is to be
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:108]: I'm not sure this detail about these particular operators is useful. 

In general NumPy arrays can't be used in torch operations and SOME NumPy operations will work with torch tensors but they don't work on them in general.

File: docs/source/notes/numpy_compatibility.rst [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+
+   "``+``", "``np.add()``", "Addition"
+   "``-``", "``np.subtract()``", "Subtraction"
+   "``*``", "``np.multiply()``", "Multiplication"
+   "``/``", "``np.divide()``", "Division"
+
+Conversions
+-----------
+
+Only a small subset of data type (``dtype``) objects defined in NumPy have an
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:145]: I would avoid editorializing judgments like "small"

File: docs/source/notes/numpy_compatibility.rst [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+    "``int$``", "``int$``", ":math:`8, 16, 32, 64`"
+    "``float$``", "``float$``", ":math:`16, 32, 64`"
+    "``complex$``", "``complex$``", ":math:`64, 128`"
+
+To ``numpy``
+^^^^^^^^^^^^
+
+The restrictions on a PyTorch tensor becoming a NumPy ``ndarray`` are:
+
+- It must be a dense tensor
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:162]: "dense" -> "strided"

File: docs/source/notes/numpy_compatibility.rst [hunk 21]
```diff
@@ -0,0 +1,388 @@
...
+- It must be on the CPU
+- It must not have a computational graph
+- It must not have the conjugate bit set
+- It must not have the negative bit set
+- It must not be a subclass
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:167]: "subclass" -> "tensor subclass"

File: docs/source/notes/numpy_compatibility.rst [hunk 22]
```diff
@@ -0,0 +1,388 @@
...
+    "``complex$``", "``complex$``", ":math:`64, 128`"
+
+To ``numpy``
+^^^^^^^^^^^^
+
+The restrictions on a PyTorch tensor becoming a NumPy ``ndarray`` are:
+
+- It must be a dense tensor
+- It must be on the CPU
+- It must not have a computational graph
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:164]: "computation graph" -> "requires grad"?

File: docs/source/notes/numpy_compatibility.rst [hunk 23]
```diff
@@ -0,0 +1,388 @@
...
+   requirement still needs to be satisfied by the user, i.e., ``requires_grad ==
+   False``
+
+From ``numpy``
+^^^^^^^^^^^^^^
+
+For a NumPy ``ndarray`` to be convertible to a PyTorch tensor:
+
+- It must have only native byte order
+- Must not be read only
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:213]: Don't we allow the conversion of non-writable arrays?

File: docs/source/notes/numpy_compatibility.rst [hunk 24]
```diff
@@ -0,0 +1,388 @@
...
+
+- :meth:`torch.from_numpy()` is guaranteed to share memory with NumPy.
+- :meth:`torch.as_tensor()` will try to stay away from copy operations, it
+  also has the effect of sharing memory. However, ``torch.as_tensor()`` has
+  slightly higher overhead as it checks and accepts other iteratable objects as
+  well, e.g. ``list`` objects.
+- :meth:`torch.from_dlpack()` called with a NumPy array (``np.version.version >=
+  1.20``) as its argument will also generate a ``torch.Tensor`` view.
+
+To obtain a **copy** of the ``ndarray`` object, and not share memory, the
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:257]: Alternatively the user can explicitly make a copy after the conversion to a tensor

File: docs/source/notes/numpy_compatibility.rst [hunk 25]
```diff
@@ -0,0 +1,388 @@
...
+   tensor([1., 2., 3.], dtype=torch.float64)
+   >>> b_torch[2] = 23
+   >>> np.array_equal(b_torch.numpy(), a_np)
+   False
+   >>> c_np = b_torch.numpy().copy() # new memory
+
+Calling NumPy Functions on PyTorch
+----------------------------------
+
+Operators aside, NumPy functions can be called on PyTorch tensors as well. This
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:280]: This is not generally true. NumPy operations don't work on CUDA tensors, for example.

I'm not even sure this is true for CPU tensors. Don't some NumPy operations call methods on the original object and our implementation of those methods is (unfortunately) divergence from NumPy's?

File: docs/source/notes/numpy_compatibility.rst [hunk 26]
```diff
@@ -0,0 +1,388 @@
...
+   ...
+   ValueError: At least one stride in the given numpy array is negative, and tensors with negative strides are not currently supported. (You can probably work around this by making a copy of your array  with array.copy().)
+   >>> b_torch = torch.from_numpy(np.ascontiguousarray(a_np[::-1]))
+   >>> b_torch
+   tensor([3, 2, 1])
+
+Conclusions
+-----------
+
+For the most part, implicit conversions between NumPy and PyTorch tensors are
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:356]: I feel like this doc is showing a lot of work to justify this statement but as a document for casual users we should probably just be clear about what we recommend/support upfront and not worry so much about proving ourselves correct

File: docs/source/notes/numpy_compatibility.rst [hunk 27]
```diff
@@ -0,0 +1,388 @@
...
+
+`NEP 13`_ and `NEP 18`_, define ``__array_ufunc__`` and ``__array_function__``
+respectively. Neither of these have been implemented in PyTorch, and since these
+mechanisms have largely been replaced by newer approaches, they are unlikely to
+be included.
+
+The Array API
+~~~~~~~~~~~~~
+
+The main source of confusion is that the existing NumPy API is far too forgiving
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:371]: Source of confusion over what?

File: docs/source/notes/numpy_compatibility.rst [hunk 28]
```diff
@@ -0,0 +1,388 @@
...
+   import torch
+
+NumPy Compatibility
+===================
+
+.. note::
+
+   **TL;DR:**
+
+   - Be explicit about obtaining and using NumPy arrays
```
Comment (mruberry) [inline docs/source/notes/numpy_compatibility.rst:15]: Having now read the document, I like being clear about converting between torch tensors and NumPy arrays. I think we should just tell people if they want to call NumPy operations on PyTorch tensors how they can do that (like creating a NumPy view of the data and running in a no grad context). We should let people know that PyTorch has a very "NumPy compatible" API and try to note the biggest divergences. We should talk about our Python Array API support (which is still in development) and let the user know if they can acquire the Python array API namespace from a torch tensor (not at this time).

File: docs/source/notes/numpy_compatibility.rst (hunk 11 above)
Comment (rgommers) [reply docs/source/notes/numpy_compatibility.rst:51]: Maybe "instead of mixing PyTorch tensors with NumPy arrays in function calls or Python's built-in operators"? I can't immediately think of any exceptions there, and it's by design that only `Tensor`, `Tensor` subclasses, and other explicitly-allowed objects like the ones with `__torch_function__` / `__torch_dispatch__` methods are accepted.

File: docs/source/notes/numpy_compatibility.rst (hunk 25 above)
Comment (rgommers) [reply docs/source/notes/numpy_compatibility.rst:280]: Yes, this is indeed the case for (IIRC) 6 reductions, like `max`/`mean`/`std`:
```python
>>> t = torch.tensor([1, 2, 3])
>>> np.std(t)
Traceback (most recent call last):
  File "<ipython-input-8-42c835b8b08a>", line 1, in <module>
    np.std(t)
  File "<__array_function__ internals>", line 180, in std
  File "/home/rgommers/anaconda3/envs/pytorch-111-nightly-jan2021/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 3613, in std
    return std(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)
TypeError: std() received an invalid combination of arguments 
```

Comment (rgommers) [reply docs/source/notes/numpy_compatibility.rst:280]: Those are the unfortunate exceptions, I think everything else does work.

File: docs/source/notes/numpy_compatibility.rst (hunk 23 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:213]: From `tensor_numpy.cpp` in `csrc/utils/`, it **is** supported but throws a warning:

```
  TORCH_WARN_ONCE(
    "The given NumPy array is not writable, and PyTorch does "
    "not support non-writable tensors. This means writing to this tensor "
    "will result in undefined behavior. "
    "You may want to copy the array to protect its data or make it writable "
    "before converting it to a tensor. This type of warning will be "
    "suppressed for the rest of this program.");
}
```

I'll reword this to strongly suggest that the array should be writable.

File: docs/source/notes/numpy_compatibility.rst (hunk 14 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:66]: Adding the other checks (conj and neg), but I was thinking it might be more pedagogical to see each requirement and the corresponding "resolution method". So perhaps:

```python
   def uncond_numpy(torch_tensor):
        if torch_tensor.device != torch.device(type="cpu"):
           torch_tensor = torch_tensor.cpu()
        if torch_tensor.is_conj() == True:
           torch_tensor = torch_tensor.resolve_conj()
        if torch_tensor.is_neg() == True:
           torch_tensor = torch_tensor.resolve_neg()
        if torch_tensor.requires_grad == False:
            return torch_tensor.numpy()
        else:
            return torch_tensor.detach().numpy()
```

File: docs/source/notes/numpy_compatibility.rst (hunk 16 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:75]: It certainly seems to be so:
```python
>>> b_torch = torch.ones(3)
>>> a_np = np.ones(3)
>>> torch.cross(b_torch, a_np)
>>> torch.add(b_torch, a_np)
```
Both of these will raise `argument 'other' (position 2) must be Tensor, not numpy.ndarray`

File: docs/source/notes/numpy_compatibility.rst (hunk 25 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:285]: Rewording and adding exceptions

File: docs/source/notes/numpy_compatibility.rst (hunk 10 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:39]: Adding a footnote and linking to the `autograd` intro.

File: docs/source/notes/numpy_compatibility.rst (hunk 27 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:376]: Rewording.

File: docs/source/notes/numpy_compatibility.rst (hunk 12 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:54]: Removed.

File: docs/source/notes/numpy_compatibility.rst (hunk 15 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:35]: Added a section on views along with more examples of the conversions.

File: docs/source/notes/numpy_compatibility.rst (hunk 18 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:112]: Abstracted into an optional dropdown.

File: docs/source/notes/numpy_compatibility.rst (hunk 17 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:77]: Reworded

File: docs/source/notes/numpy_compatibility.rst (hunk 16 above)
Comment (HaoZeke) [reply docs/source/notes/numpy_compatibility.rst:75]: I've tried to reword this but I'm not sure it is as clear as possible.

**Comment by HaoZeke** [comment]
@rgommers and @mruberry, this has now been revised in light of https://github.com/pytorch/pytorch/pull/59790 being merged.

Does it read alright?

**Review by rgommers — CHANGES_REQUESTED** [review]
This now recommends in a couple of places to use `.numpy(force=True)`, which is not a good idea. That feature is something that is meant for a use case like described in one of the issues: Napari wants to provide generic support for array/tensor libraries, and as a visualization library doesn't need to worry about performance as much or the implications of implicit device transfer or `.detach()`. For the average user on the other hand, we don't want to teach such bad habits and sweep all the details under the rug.

This still doesn't read well, it still feels like a long list of weird corner cases. Concrete example of that: there's a red warning box near the top with _"Do not use `id` to check for shared memory"_, which just feels like a big distraction to the story (I have never seen any user do this in practice).

Most of the content is there, it just needs to be polished up with a focus on teaching, and hiding all the corner cases as notes in dropdown boxes instead of giving them center stage.

@HaoZeke I propose to put this on hold for now, for two reasons: (1) after multiple iterations it still feels off the mark, and (2) there are higher-prio things you have started on (primtorch). With a few hours to refactor this I think it's mergable, but that's probably something @mruberry or I should do.

@mruberry's main review comment from before was:

_"Having now read the document, I like being clear about converting between torch tensors and NumPy arrays. I think we should just tell people if they want to call NumPy operations on PyTorch tensors how they can do that (like creating a NumPy view of the data and running in a no grad context). We should let people know that PyTorch has a very "NumPy compatible" API and try to note the biggest divergences. We should talk about our Python Array API support (which is still in development) and let the user know if they can acquire the Python array API namespace from a torch tensor (not at this time)."_

It doesn't feel like that was addressed.

**Comment by rgommers** [comment]
I will close this for now. The content here can be reused in case someone is picking up the linked doc issue. 

Thanks for the effort @HaoZeke and @mruberry
